File: 3/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        approach
        keep 2 pointers, start and end
        end is in loop
        keep a running count of chars
        if has repeats, move left
        get max length
        abcabcbb
        i | ch | slow | max_len | chars
        0 | a  | 0    | 0       | {a: 1}
        1 | b  | 0    | 2       | {a: 1, b: 1}
        2 | c  | 0    | 3       | {a: 1, b: 1, c: 1}
        3 | a  | 0    | 3       | {a: 2, b: 1, c: 1}
        3 | a  | 1    | 3       | {a: 1, b: 1, c: 1}
        """
        chars = {}
        slow = 0
        max_len = 0
        n = len(s)
        for i in range(n):
            ch = s[i]
            if ch not in chars:
                chars[ch] = 0
            chars[ch] += 1
            while chars[ch] > 1:
                chars[s[slow]] -= 1
                slow += 1
            max_len = max(max_len, i - slow + 1)
        return max_len

File: 3/main.ts
function lengthOfLongestSubstring(s: string): number {
  let lengthOfLongestSubstring = 0;
  let leftPointer = -1;
  let lettersInWindow = new Map();
  for (let rightPointer = 0; rightPointer < s.length; rightPointer++) {
    let currentChar = s[rightPointer];
    if (
      lettersInWindow.has(currentChar) &&
      lettersInWindow.get(currentChar) > leftPointer
    ) {
      leftPointer = lettersInWindow.get(currentChar);
      lengthOfLongestSubstring = Math.max(
        lengthOfLongestSubstring,
        rightPointer - leftPointer
      );
    } else {
      lengthOfLongestSubstring = Math.max(
        lengthOfLongestSubstring,
        rightPointer - leftPointer + 1
      );
    }
    lettersInWindow.set(currentChar, rightPointer);
  }
  return lengthOfLongestSubstring;
}

// approach
// abcabcbb -> 3
// 0
// if (s[i] not in lettersInWindow)
// lengthOfCurrentWindow++
// if (lengthOfCurrentWindow > lengthOfLongestSubstring)
// lengthOflongestSubstring = lengthOfCurrentWindow
// lettersInWindow.set(s[i], i)
// else if (s[i] in lettersInWindow)
// lengthOfCurrentWindow = i - lettersInWindow[s[i]]
// lettersInWindow[s[i]] = i

console.log(lengthOfLongestSubstring("tmmzuxt"));
